### PR TITLE
privilege, executor: add `SET ROLE` and `CURRENT_ROLE` support

### PR DESCRIPTION
**Pull request by imtbkcat**
### What problem does this PR solve? 
support active role and `SET ROLE`, `CURRENT_ROLE` function. 

### What is changed and how it works?
Create an graph data structure to find relationship between role and user quickly, which is always update as other privilege tables. It will load `mysql.role_edges` table, and convert relationship to an graph.

When we need to active some roles for current session, we need to check whether these roles  has been granted for current user. `RoleGraph` can finish this task quickly.

`SET ROLE` is just for set active role for current session, more detail: https://dev.mysql.com/doc/refman/8.0/en/set-role.html

Because `set default role` is not support yet. some gramma will be added soon.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported variable/fields change

Side effects

 - Increased code complexity

### CODE REVIEW

File: privilege/privileges/cache.go [hunk 1]
```diff
@@ -16,6 +16,7 @@ package privileges
 import (
 	"context"
 	"fmt"
+	"github.com/pingcap/parser/auth"
```
Comment (jackysp) [inline privilege/privileges/cache.go]: Move this line to the third party libs part.

File: expression/builtin_info.go [hunk 2]
```diff
@@ -156,6 +157,50 @@ func (b *builtinCurrentUserSig) evalString(row chunk.Row) (string, bool, error)
...
+
+func (b *builtinCurrentRoleSig) Clone() builtinFunc {
+	newSig := &builtinCurrentRoleSig{}
+	newSig.cloneFrom(&b.baseBuiltinFunc)
+	return newSig
+}
+
+// evalString evals a builtinCurrentUserSig.
+// See https://dev.mysql.com/doc/refman/5.7/en/information-functions.html#function_current-user
+func (b *builtinCurrentRoleSig) evalString(row chunk.Row) (string, bool, error) {
```
Comment (jackysp) [inline expression/builtin_info.go]: It is better to add some test cases for this built-in function.

Comment (imtbkcat) [reply expression/builtin_info.go]: ok

File: executor/simple.go [hunk 3]
```diff
@@ -80,11 +80,36 @@ func (e *SimpleExec) Next(ctx context.Context, req *chunk.RecordBatch) (err erro
...
+	roleList := make([]*auth.RoleIdentity, 0, 10)
+	for _, v := range checkDup {
+		roleList = append(roleList, v)
+	}
+
+	checker := privilege.GetPrivilegeManager(e.ctx)
+	ok, roleName := checker.ActiveRoles(e.ctx, roleList)
+	if !ok {
+		errCode := mysql.ErrRoleNotGranted
+		errRoleNotGranted := terror.ClassPrivilege.New(terror.ErrCode(errCode), mysql.MySQLErrName[uint16(errCode)])
```
Comment (tiancaiamao) [inline executor/simple.go]: define this error in the file `executor/errors.go`

File: privilege/privileges/cache.go [hunk 4]
```diff
@@ -104,12 +105,45 @@ type columnsPrivRecord struct {
...
+func (g RoleGraphEdgesTable) Find(user, host string) bool {
+	if host == "" {
+		host = "%"
+	}
+	key := user + "@" + host
+	if g.roleList == nil {
+		return false
+	}
+	_, ok := g.roleList[key]
+	if ok {
```
Comment (tiancaiamao) [inline privilege/privileges/cache.go]: return ok

File: privilege/privileges/cache.go [hunk 5]
```diff
@@ -104,12 +105,45 @@ type columnsPrivRecord struct {
...
 // MySQLPrivilege is the in-memory cache of mysql privilege tables.
 type MySQLPrivilege struct {
 	User        []UserRecord
 	DB          []dbRecord
 	TablesPriv  []tablesPrivRecord
 	ColumnsPriv []columnsPrivRecord
+	RoleGraph   map[string]RoleGraphEdgesTable
+}
+
+// FindRole is used to detect whether this is an edge between user and role
```
Comment (tiancaiamao) [inline privilege/privileges/cache.go]: s/this is/there is

File: privilege/privileges/cache.go [hunk 6]
```diff
@@ -104,12 +105,45 @@ type columnsPrivRecord struct {
 	patTypes []byte
 }
 
+// RoleGraphEdgesTable is used to cache relationship between and role.
+type RoleGraphEdgesTable struct {
```
Comment (tiancaiamao) [inline privilege/privileges/cache.go]: s/RoleGraphEdgesTable/roleGraphEdgesTable

File: privilege/privileges/cache.go [hunk 7]
```diff
@@ -381,6 +433,31 @@ func (p *MySQLPrivilege) decodeTablesPrivTableRow(row chunk.Row, fs []*ast.Resul
...
+			fromUser = row.GetString(i)
+		case f.ColumnAsName.L == "to_host":
+			toHost = row.GetString(i)
+		case f.ColumnAsName.L == "to_user":
+			toUser = row.GetString(i)
+		}
+	}
+	fromKey := fromUser + "@" + fromHost
+	toKey := toUser + "@" + toHost
+	if _, ok := p.RoleGraph[toKey]; ok {
```
Comment (tiancaiamao) [inline privilege/privileges/cache.go]:
```
roleGraph, ok := p.RoleGraph[toKey];
if !ok {
    roleGraph = RoleGraphEdgesTable{roleList: make(map[string]bool)}
    p.RoleGraph[toKey] = roleGraph
}
roleGraph.roleList[fromKey] = true
```

**Comment by imtbkcat** [comment]
PTAL @tiancaiamao

File: executor/simple.go [hunk 8]
```diff
@@ -80,11 +80,34 @@ func (e *SimpleExec) Next(ctx context.Context, req *chunk.RecordBatch) (err erro
...
 		err = e.executeDropStats(x)
+	case *ast.SetRoleStmt:
+		err = e.executeSetRole(x)
 	}
 	e.done = true
 	return errors.Trace(err)
 }
 
+func (e *SimpleExec) executeSetRole(s *ast.SetRoleStmt) error {
+	checkDup := make(map[string]*auth.RoleIdentity)
```
Comment (tiancaiamao) [inline executor/simple.go]: `make(map[string]*auth.RoleIdentity, len(s.RoleList))`

File: go.mod [hunk 9]
```diff
@@ -23,7 +23,7 @@ require (
 	github.com/go-sql-driver/mysql v0.0.0-20170715192408-3955978caca4
 	github.com/gogo/protobuf v1.2.0 // indirect
 	github.com/golang/groupcache v0.0.0-20181024230925-c65c006176ff // indirect
-	github.com/golang/protobuf v1.2.0
+	github.com/golang/protobuf v1.3.0
```
Comment (tiancaiamao) [inline go.mod]: Why protobuf is changed here?

File: privilege/privileges/cache.go [hunk 10]
```diff
@@ -104,12 +105,42 @@ type columnsPrivRecord struct {
 	patTypes []byte
 }
 
+// RoleGraphEdgesTable is used to cache relationship between and role.
+type roleGraphEdgesTable struct {
+	roleList map[string]bool
```
Comment (tiancaiamao) [inline privilege/privileges/cache.go]: It's better to use `roleList map[string]struct{}` here

```
roleList map[string]struct{}

roleList["xx"] = struct{}{}

if _, ok := roleList["xx"] {
    ...
}
```

**Comment by tiancaiamao** [comment]
LGTM

**Comment by imtbkcat** [comment]
PTAL @jackysp

**Pull request merged**